### PR TITLE
Add LIP "Introduce sparse Merkle trees"

### PR DESCRIPTION
**Pull request by ricott1**

### CODE REVIEW

**Review by IkerAlus — APPROVED** [review]
All looks good to me, great work.

**Review by shuse2 — APPROVED** [review]

File: proposals/lip-Introduce-sparse-Merkle-trees.md [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+leafHash(msg) = hash(LEAF_PREFIX | msg)
+branchHash(msg) = hash(BRANCH_PREFIX | msg),
+```
+
+where `LEAF_PREFIX` and `BRANCH_PREFIX` are two constants set to:
+
+```java
+LEAF_PREFIX = 0x00
+BRANCH_PREFIX = 0x01.
+```
```
Comment (janhack) [inline proposals/lip-Introduce-sparse-Merkle-trees.md:129]: As we have the table of constants, lines 123-129 are redundant.

File: proposals/lip-Introduce-sparse-Merkle-trees.md [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+## Specification
+
+For the rest of this proposal, we define the following constants:
+
+| Name          | Type    | Value       |
+| ------------- |---------| ------------|
+| EMPTY_HASH    | bytes   | SHA-256("") |
+| LEAF_PREFIX   | bytes   | 0x00        |
+| BRANCH_PREFIX | bytes   | 0x01        |
+| KEY_LENGTH    | integer | Set when initializing the data structure |
```
Comment (janhack) [inline proposals/lip-Introduce-sparse-Merkle-trees.md:95]: Here an below it is not clear what unit is used for the key length.

Maybe the parameter name could be `KEY_LENGTH_BYTES` or something similar. An for the description here we could as "Length of the key in bytes."

File: proposals/lip-Introduce-sparse-Merkle-trees.md [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+    let queryBitmap be an empty string
+    let currentNode be the tree root node 
+    if currentNode is defaultEmptyNode:
+        let response be a data structure with:
+            {key=queryKey, 
+             value=empty buffer,
+             binaryBitmap=empty string,
+             siblingHashes=siblingHashes,
+             ancestorHashes=ancestorHashes}
+    return response
```
Comment (janhack) [inline proposals/lip-Introduce-sparse-Merkle-trees.md:333]: The indentation above is incorrect.

File: proposals/lip-Introduce-sparse-Merkle-trees.md [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+    if length of queryKeys != length of proof.queries:
+        return false
+    for each query q in proof.queries:
+        let k be the corresponding key in queryKeys (the element of queryKeys with same index)
+        //q is an inclusion proof for k or a default empty node 
+        if q.key == k:
+            continue
+        //q is an inclusion proof for another leaf node
+        let binaryQueryBitmap = binaryExpansion(q.bitmap) without leading zeros
+        let binaryQueryKey=binaryExpansion(q.key)
```
Comment (janhack) [inline proposals/lip-Introduce-sparse-Merkle-trees.md:534]: As query key I would understand the key provided in the query. Maybe proof key would be a better name here?

**Review by janhack — CHANGES_REQUESTED** [review]
I left a few minor comments.

File: proposals/lip-Introduce-sparse-Merkle-trees.md (hunk 3 above)
Comment (ricott1) [reply proposals/lip-Introduce-sparse-Merkle-trees.md:333]: fixed

File: proposals/lip-Introduce-sparse-Merkle-trees.md [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+        *   `value`: A byte value giving the value of the node associated with `key.`
+
+The protocol for an inclusion proof is similar to the [one specified for regular Merkle trees](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0031.md#proof-of-inclusion-protocol):
+
+*   Verifier: The Verifier knows `merkleRoot` and sends the Prover an array of keys, `queryKeys`, for which they wish to obtain a proof-of-inclusion;
+*   Prover:
+    *   The Prover finds the leaf node corresponding to each key `k` in `queryKeys` and assigns to each a `query` object with a key, a bitmap, and a value. The bitmap indicates the presence of default empty nodes along the path to the Merkle root. If the node exists, `query.key` is set to `k` and `query.value` to the node value; if the node does not exist, the Verifier returns an inclusion proof for another leaf or a default empty node in the path of the `k`. In this case, `query.key` and `query.value` are set to the leaf key and leaf value or to `k` and empty value, respectively.
```
Comment (Incede) [inline proposals/lip-Introduce-sparse-Merkle-trees.md:80]:
```suggestion
    *   The Prover finds the leaf node corresponding to each key `k` in `queryKeys` and assigns to each a `query` object with a key, a bitmap, and a value. The bitmap indicates the presence of default empty nodes along the path to the Merkle root. If the node exists, `query.key` is set to `k` and `query.value` to the node value; if the node does not exist, the Prover returns an inclusion proof for another leaf or a default empty node in the path of the `k`. In this case, `query.key` and `query.value` are set to the leaf key and leaf value or to `k` and empty value, respectively.  
```

**Review by Incede — CHANGES_REQUESTED** [review]

File: proposals/lip-Introduce-sparse-Merkle-trees.md (hunk 4 above)
Comment (janhack) [reply proposals/lip-Introduce-sparse-Merkle-trees.md]: Looks good. That is more clear.

**Review by janhack — APPROVED** [review]

File: proposals/lip-Introduce-sparse-Merkle-trees.md (hunk 5 above)
Comment (ricott1) [reply proposals/lip-Introduce-sparse-Merkle-trees.md:80]: nice catch! thanks

**Review by Incede — APPROVED** [review]

**Review by karmacoma — APPROVED** [review]

**Pull request merged**